efi/preinstall/checks_context.go

```diff
@@ -0,0 +1,585 @@
+// should match the flags later passed to this function when creating a PCR profile. The PCRs that are
+// determined to be required to build the profile using [WithAutoTCGPCRProfile] will be made mandatory
+// automatically by this function by passing the relevant flags to [RunChecks].
+//
+// As [PCRProfileOptionsFlags] is intended to provide some limited amount of user customisation, this
+// API [RunChecksContext] should be executed again with the new set of flags, should the user wish to
+// change them. In this case, the caller should pass the PostInstallChecks flag as an initial flag.
+//
+// There is no need for the caller to supply any of these *SupportRequired flags as the initial flags,
+// and this may have the effect of limiting the number of devices which pass the checks.
+func NewRunChecksContext(initialFlags CheckFlags, loadedImages []secboot_efi.Image, profileOpts PCRProfileOptionsFlags) *RunChecksContext {
+	return &RunChecksContext{
+		env:          runChecksEnv,
+		flags:        initialFlags,
+		loadedImages: loadedImages,
+		profileOpts:  profileOpts,
+		// Populate actions that are always available or available by default
+		// unless we discover that they aren't later on.
+		availableActions: map[Action]bool{
+			ActionNone:               true,
+			ActionReboot:             true,
+			ActionShutdown:           true,
+			ActionRebootToFWSettings: true,
+			ActionContactOEM:         true,
+			ActionContactOSVendor:    true,
+		},
+		expectedActions: []Action{ActionNone},
+	}
+}
+
+// testActionAvailable will perform some tests in order to determine whether
+// the specified action is available.
+func (c *RunChecksContext) testActionAvailable(action Action) error {
+	available := false
+
+	switch action {
+	// TODO: Populate with actions to test as we add them later on.
+	}
+
+	c.availableActions[action] = available
+	return nil
+}
+
+// filterUnavailableActions will filter out any actions in the supplied slice
+// that are unavailable, and return a new slice containing only actions that
+// are available.
+func (c *RunChecksContext) filterUnavailableActions(actions []Action) (out []Action, err error) {
+	for _, action := range actions {
+		available, tested := c.availableActions[action]
+		switch {
+		case !tested:
+			if err := c.testActionAvailable(action); err != nil {
+				return nil, fmt.Errorf("cannot test whether action %q is available: %w", action, err)
+			}
+			if available := c.availableActions[action]; available {
+				out = append(out, action)
+			}
+		case available:
+			out = append(out, action)
+		}
+	}
+
+	return out, nil
+}
+
+// isActionExpected determines if the supplied action is an expected
+// response to the last call to [RunChecksContext.Run].
+func (c *RunChecksContext) isActionExpected(action Action) bool {
+	for _, expected := range c.expectedActions {
+		if expected == action {
+			return true
+		}
+	}
+	return false
```

Comment:
tests don't seem to hit the false case

Reply:
I can add a test for this.

Reply:
I've added a test case with an invalid action name now.